crates/protocol/src/block_info.rs

```diff
@@ -0,0 +1,532 @@
+
+        // With the regolith hardfork, system transactions were deprecated, and we allocate
+        // a constant amount of gas for special transactions like L1 block info.
+        if rollup_config.is_regolith_active(l2_block_time) {
+            deposit_tx.is_system_transaction = false;
+            deposit_tx.gas_limit = REGOLITH_SYSTEM_TX_GAS;
+        }
+
+        Ok((l1_info, OpTxEnvelope::Deposit(deposit_tx)))
+    }
+
+    /// Decodes the [L1BlockInfoEcotone] object from ethereum transaction calldata.
+    pub fn decode_calldata(r: &[u8]) -> Result<Self> {
+        let selector = r[0..4].try_into().expect("Failed to convert 4byte slice to array");
+        match selector {
+            L1_INFO_TX_SELECTOR_BEDROCK => {
+                Ok(Self::Bedrock(L1BlockInfoBedrock::decode_calldata(r)?))
+            }
+            L1_INFO_TX_SELECTOR_ECOTONE => {
+                Ok(Self::Ecotone(L1BlockInfoEcotone::decode_calldata(r)?))
+            }
+            _ => anyhow::bail!("Unreachable case; Invalid L1 info transaction selector."),
+        }
+    }
+
+    /// Encodes the [L1BlockInfoTx] object into Ethereum transaction calldata.
+    pub fn encode_calldata(&self) -> Bytes {
+        match self {
+            Self::Bedrock(bedrock_tx) => bedrock_tx.encode_calldata(),
+            Self::Ecotone(ecotone_tx) => ecotone_tx.encode_calldata(),
+        }
+    }
+
+    /// Returns the L1 [BlockID] for the info transaction.
+    pub fn id(&self) -> BlockID {
+        match self {
+            Self::Ecotone(L1BlockInfoEcotone { number, block_hash, .. }) => {
+                BlockID { number: *number, hash: *block_hash }
+            }
+            Self::Bedrock(L1BlockInfoBedrock { number, block_hash, .. }) => {
+                BlockID { number: *number, hash: *block_hash }
+            }
+        }
+    }
+
+    /// Returns the L1 fee overhead for the info transaction. After ecotone, this value is ignored.
+    pub fn l1_fee_overhead(&self) -> U256 {
+        match self {
+            Self::Bedrock(L1BlockInfoBedrock { l1_fee_overhead, .. }) => *l1_fee_overhead,
+            Self::Ecotone(_) => U256::ZERO,
+        }
+    }
+
+    /// Returns the batcher address for the info transaction
+    pub fn batcher_address(&self) -> Address {
+        match self {
+            Self::Bedrock(L1BlockInfoBedrock { batcher_address, .. }) => *batcher_address,
+            Self::Ecotone(L1BlockInfoEcotone { batcher_address, .. }) => *batcher_address,
+        }
+    }
+
+    /// Returns the sequence number for the info transaction
+    pub fn sequence_number(&self) -> u64 {
+        match self {
+            Self::Bedrock(L1BlockInfoBedrock { sequence_number, .. }) => *sequence_number,
+            Self::Ecotone(L1BlockInfoEcotone { sequence_number, .. }) => *sequence_number,
+        }
+    }
+}
+
+impl L1BlockInfoBedrock {
+    /// Encodes the [L1BlockInfoBedrock] object into Ethereum transaction calldata.
+    pub fn encode_calldata(&self) -> Bytes {
```

Comment:
We should probably just implement the [Decodable](https://docs.rs/alloy-rlp/latest/alloy_rlp/trait.Decodable.html) and [Encodable](https://docs.rs/alloy-rlp/latest/alloy_rlp/trait.Encodable.html) traits here from `alloy_rlp` instead of custom methods.

Reply:
Curious if you have thoughts @clabby